### PR TITLE
Add --measure-time flag to libtest

### PR DESCRIPTION
**Pull request by popzxc**
## Overview

This PR introduces a new option flag for `rust` `test` command: `measure-time`. With this flag applied, `rustc` will measure the execution time of each test and print it out (into file log or if `pretty` formatter is chosen).

`--measure-time` is intended for a single-threaded run, so using with `--test-threads` set to any value other than `1` will result in an error.

Time is measured only when flag is applied, thus there is no overhead for common runs.

In the output, time is written in red if test duration time >= `RUST_TEST_TIME_CRITICAL` (defaults to 1000 milliseconds), in yellow if test duration time >= `RUST_TEST_TIME_WARN` (defaults to 500 milliseconds), otherwise default formatting is applied.

rustc_timed_tests

## Rationale

Usually, big Rust projects have a lot of tests. And tests are not always written carefully. As a result, if some poorly written test gets merged into a big project, it will be difficult to find that test, because there is no tool for that. `libtest` checks for timed out tests, but the constant `TEST_WARN_TIMEOUT_S` is not configurable, and it's not really an instrument to detect not optimized tests.

This PR allows developers to inspect their codebase to detect long-running tests and optimize/rewrite/remove them.

## Backward compatibility

This PR is fully backward compatible, it does not introduce any breaking changes.

r? @Centril

### CODE REVIEW

**Comment by fbstj** [comment]
is this different to #64663

**Comment by popzxc** [comment]
Oh... Well, I started implementing this before that PR was submitted :c

However, sort of yes.

This PR does (and linked does not):
- verifies that only 1 test thread is used (for tests running in parallel results may be inprecise).
- highlights tests with big execution time.
- does not report time for failed tests (failed test usually not executed fully, so it's not quite correct to report its execution time).
- does not introduce overhead for common runs (`Instant::now` [may require](https://doc.rust-lang.org/src/std/time.rs.html#189) `Mutex` lock in several cases.
- a bit more configurable.

File: src/libtest/lib.rs
Comment (Centril) [inline src/libtest/lib.rs]:
```suggestion
        .optflag("", "measure-time", "Measure test execution time")
```

File: src/libtest/lib.rs
Comment (Centril) [inline src/libtest/lib.rs]: A guard would be better here.

File: src/libtest/lib.rs
Comment (Centril) [inline src/libtest/lib.rs]:
```suggestion
                    return Some(Err("The \"measure-time\" flag is only compatible with one thread".into()));
```

File: src/libtest/lib.rs
Comment (Centril) [inline src/libtest/lib.rs]: Same; use a guard.

File: src/libtest/lib.rs
Comment (Centril) [inline src/libtest/lib.rs]: Let binding here seems unnecessary?

File: src/libtest/lib.rs
Comment (Centril) [inline src/libtest/lib.rs]: Inline this `let`, use `.parse::<usize>()`. Also, `.ok()` is redundant, just match on `Ok(n)` instead.

File: src/libtest/lib.rs
Comment (Centril) [inline src/libtest/lib.rs]: This could be a doc comment.

File: src/libtest/lib.rs
Comment (Centril) [inline src/libtest/lib.rs]: Add some doc comments to this one.

File: src/libtest/lib.rs
Comment (Centril) [inline src/libtest/lib.rs]: De-duplicate this with a closure or function taking `key: &str, fallback: u64`.

File: src/libtest/lib.rs
Comment (Centril) [inline src/libtest/lib.rs]: Why unpack this just to pack it again?

File: src/libtest/lib.rs
Comment (Centril) [inline src/libtest/lib.rs]:
```suggestion
    duration: Option<Duration>,
```

File: src/libtest/tests.rs
Comment (Centril) [inline src/libtest/tests.rs]:
```suggestion
    // Provide `--measure-time` and `--test` flags.
```

File: src/libtest/tests.rs
Comment (Centril) [inline src/libtest/tests.rs]:
```suggestion
    // Provide `--measure-time` flag without `--test` flag and with `--bench` flag.
```

File: src/libtest/tests.rs
Comment (Centril) [inline src/libtest/tests.rs]:
```suggestion
    // Provide `--measure-time` flag with custom `--test-threads` flag.
```

**Comment by Centril** [comment]
Seems reasonable -- should this be the default perhaps?

Not sure if this is feature gated or not, but I think we'd want to gate the behavior initially.

cc @alexcrichton @gnzlbg @killercup

**Comment by gnzlbg** [comment]
See also #64663 which will probably be merged. Maybe this should build on that ?

**Comment by Centril** [comment]
Seems like a good idea to do so yeah.

**Comment by gnzlbg** [comment]
> --measure-time is intended for a single-threaded run, so using with --test-threads set to any value other than 1 will result in an error.

Why is this? It should be possible to measure the time it takes to run a test independently of whether the tests are run in parallel or not.

> n the output, time is written in red if test duration time >= RUST_TEST_TIME_CRITICAL (defaults to 1000 milliseconds), in yellow if test duration time >= RUST_TEST_TIME_WARN (defaults to 500 milliseconds), otherwise default formatting is applied.
>
> [...] if some poorly written test gets merged into a big project, it will be difficult to find that test, because there is no tool for that. libtest checks for timed out tests, but the constant TEST_WARN_TIMEOUT_S is not configurable, and it's not really an instrument to detect not optimized tests.

With #64663, the test times are written to libtest json output - so if this is a real issue for real projects, that should allow them to write a simple tool that parses that and rejects test with longer runtimes using whatever metrics they want. That feels like a better solution to this problem than trying to solve this on the libtest side.

**Comment by popzxc** [comment]
Hm... Ok, all the comments are seem reasonable.
Will it be ok then if this PR will be closed, and once #64663 gets merged, I'll create a new PR based on updated libtest with all the review comments fixed?

@gnzlbg, regarding test threads -- ok, then maybe use a following approach: If not of the `--test-threads` and `RUST_TEST_THREADS` then tests will be runned in one thread, otherwise a provided amount of threads will be used?

**Comment by gnzlbg** [comment]
>  ok, then maybe use a following approach: If not of the --test-threads and RUST_TEST_THREADS then tests will be runned in one thread, otherwise a provided amount of threads will be used?

Why do you want to force tests to run in one thread when reporting the run-time ?

**Comment by popzxc** [comment]
> Why do you want to force tests to run in one thread when reporting the run-time ?

Imagine some resource acquired only synchronously (e.g. resource can be reached only after locking a static `Mutex`). In that hypothetical case, if tests are run in multiple threads (lets take 4) it's possible with 3 quick tests and 1 very long one to get a situation when all 4 tests are reported to be long (long test locked the mutex first, 3 other ones waited for it, as a result, 3 tests were blocked until the long one is completed). You can see the [gist](https://gist.github.com/popzxc/b30f5a9f1763cb31d594524e24717ca4) for the actual example.

I understand that it's not a common case, but still a possible one, and I believe that compiler tools should avoid ambiguity when possible.

Overall, if developers are sure that it's ok to run tests in parallel, they always will be able to provide a compiler flag for it.

Or am I missing something here?

**Comment by gnzlbg** [comment]
libtest runs tests in parallel by default and it offers an option to choose the number of threads.
A libtest option for reporting how long does it take for a unit test to run should just that, and is orthogonal to the number of threads that are used.

> In that hypothetical case, if tests are run in multiple threads (lets take 4) it's possible with 3 quick tests and 1 very long one to get a situation when all 4 tests are reported to be long (long test locked the mutex first, 3 other ones waited for it, as a result, 3 tests were blocked until the long one is completed) [...] I believe that compiler tools should avoid ambiguity when possible.

That's not an "ambiguity", that's printing out run-times that reflect reality. Changing the number of threads used by libtest silently when a user tries to debug these issues would be super confusing: "My tests run very slowly, let's print the run-times... oh when I print the run-times they run fast, WTF?!".

**Comment by popzxc** [comment]
Yeah, but when user notices that tests run very slow and decide to measure time, it will get N tests being reported as slow, when only the one of them is slow. Thus this command won't be helping at all.

IMHO, the purpose of that subcommand is to *help* user *find* the slow tests, and not just "report run-times that reflect reality".

**Comment by gnzlbg** [comment]
> Yeah, but when user notices that tests run very slow and decide to measure time, he will get N tests being reported as slow, when only the one of them is slow in fact.

This is incorrect. If a test is reported as taking N seconds to run, then it is because it took N seconds to run. That's objective and correct data.

Whether a tests takes N seconds to run because it does a lot of computation, or because its waiting doing nothing because some other tests holds a mutex lock, is a different issue.

It appears that what you want is a libtest feature that lets you know which test is making other tests run slowly. Silently messing with the number of threads doesn't achieve that:

```rust
static COUNT: AtomicUsize = AtomicUsize::new(0);
#[test] fn foo() { for i in 0..COUNT.load(SeqCst) { println!("!"); } }
#[test] fn bar() { COUNT.write(usize::max_value(), SeqCst); }
```

Here, even when using one thread, if `bar` runs before `foo`, `foo` will take a long time to run. Changing the number of threads doesn't reveal the problem.

**Comment by gnzlbg** [comment]
> IMHO, the purpose of that subcommand is to help user find the slow tests, and not just "report run-times that reflect reality".

The purpose of the other PR is to report unit test run-times. What users do with that data, is up to them.

If your intent is to solve some other problem (adding some sort of functionality to libtest that points users to the tests that are causing other tests to run slowly), then it is probably better to open an issue to discuss that. The top post in this PR does not mention anything about this.

**Comment by popzxc** [comment]
OK, you're right. I will left test threads as they are then.

**Comment by jakoschiko** [comment]
Some thoughts:

I like colored execution times, but I don't like default values. The proposed default values (500ms warn, 1000ms critical) don't fit everyone's use case. For example property based tests that take several seconds are not unusual. Running the tests with `--measure-time` and seeing only red execution times would confuse me except I had chosen the thresholds myself.

Alternatives:
- Do not enable colored execution times by default. If the user needs it, he must configure the thresholds.
- Do not implement the feature at all. If a user really needs it, he can use something like grep:
  ```
  cargo test -- --gimme-time | egrep --color -z '[1-9]{1,}[0-9]{0,}\.[0-9]{3}s'
  ```

**Comment by wesleywiser** [comment]
@Centril I don't really know anything about `libtest`. Are you assigning me for some reason specifically or just because you need a reviewer?

**Comment by Centril** [comment]
@wesleywiser Basically I did the code review, but I'm not best placed to judge whether the motivation of the PR is enough to land / account for the discussion. @varkor thought you'd be well placed as you've dealt with profiling and such things.

**Comment by wesleywiser** [comment]
Gotcha, that's fine.

**Review by wesleywiser — CHANGES_REQUESTED** [review]
Overall, I think this is a good feature addition and I don't have concerns about the overall approach. Here's the changes I'd like to see in order to land this PR:

- Suggested changes from the other reviewers should be done.
- I agree with @gnzlbg that we shouldn't be messing with the number of test threads. While the motivation is good, I think users will potentially be more confused if we don't replicate the original test environment as closely as possible when this flag is turned on. 
- Since #64663 was landed behind an unstable flag, I think this option should also be behind the unstable flag.

I think including default values for the warning and critical times is a good idea but I think we should use different values for unit tests vs integration tests. Unit tests are supposed to be very, very fast so for those, I'd recommend 100ms as critical, 50ms as warn. For integration tests, I think the current values you have are ok. If detecting what kind of test is running is complicated, then feel free to ignore this. (Note: the [Cargo book](https://doc.rust-lang.org/cargo/guide/tests.html) says that tests in `src/*` are unit tests and tests in `test/*` are integration tests.)

I'd also like to see some documentation for this flag and its configuration options in the [Unstable Features page of the Cargo book](https://doc.rust-lang.org/cargo/reference/unstable.html).

When those three main issues are addressed, I'll happily approve this!

**Comment by popzxc** [comment]
Sure, I'll perform the required changes as soon as I'll have time for this (hopefully this or next week).

**Comment by wesleywiser** [comment]
@popzxc Sounds good! If you decide to close this PR and open another, feel free to include 

`r? @wesleywiser` 

in the PR body so that I'll be assigned to it since I have the context for this.

**Comment by gnzlbg** [comment]
>  If detecting what kind of test is running is complicated, then feel free to ignore this.

I don't think we have an easy way to detect this right now (IIRC, in both cases, the only thing that libtest "sees" is a list of `#[test] #item`), It might be better to not block the rest of this PR on extending, that, so I think it would make sense to leave that for a subsequent PR.

**Pull request closed without merging**
